### PR TITLE
numeric: T4669: extend numeric.ml for inverted and range values

### PR DESCRIPTION
**Pull request by jestabro**
This is an initial draft of requested extensions to numeric.ml:

(1) allow passing an interval value to --range:
      numeric --range 0-10 2-5    (passes)
      numeric --range 0-10 7-12  (fails)

(2) define an inversion operator for --range:
      numeric --range 0-10 !12  (passes)
      numeric --range 0-10 !5    (fails)

### CODE REVIEW

File: src/numeric.ml [hunk 1]
```diff
@@ -47,15 +62,19 @@ let is_relative value =
...
-      failwith "Value is not a relative increment/decrement"
-    else number_string_from_relative s
-  else s
+    match t with
+    | Number_s s ->
+        if not (is_relative s) then
+          failwith "Value is not a relative increment/decrement"
+        else Number_s (number_string_drop_modifier s)
+    | Range_s _ ->
+        failwith "increment/decrement does not apply to a range argument"
```
Comment (dmbaturin) [inline src/numeric.ml:76]: I think we may want to use different error codes for validation failures and usage failures. The idea is that if there's a bug in a command definition, it shouldn't look like a validation failure, but rather the user should be told to report a bug.

For example, exit with 1 on validation errors and with 2 on internal errors.

**Review by dmbaturin — CHANGES_REQUESTED** [review]
Regarding the inverted range: I believe the syntax of "not in range" checks should be `--not-range 1-10 20` rather than  `--range 1-10 !20`.

Here's why:
* `--not-range` only needs another option in the argparse spec, while `!` requires custom parsing logic.
* `--not-range` and `--range` are visually distinct and their meanings are obvious, while an exclamation mark is easy to overlook.
* `--not-range` will show up in the help message automatically, so anyone can easily find out how to negate a range from `numeric --help`.

**Comment by jestabro** [comment]
Force pushed update for task number and description; requested changes will be addressed in subsequent update.

File: src/numeric.ml (hunk 1 above)
Comment (c-po) [reply src/numeric.ml:76]: How could we strip the `!` from VyOS CLI when passed to the validator?

Comment (jestabro) [reply src/numeric.ml:76]: I think Daniil's suggestion of using --not-range instead of parsing the input is convincing, so it would be a non-issue.

Comment (dmbaturin) [reply src/numeric.ml:76]: I think I see where the idea to use `!` comes from: from existing options like firewall ports! I believe we should change that CLI to multi nodes and separate monolithic options into, say, `port number`, `port range` and `port exclude-range`. Then we can properly validate each port without parsing the argument string.

Comment (c-po) [reply src/numeric.ml:76]: I am actually working on the `<multi/>` node port implementation for https://github.com/vyos/vyos-1x/pull/1502 which brought this up.

I am fine with `port` and `exclude` port options. We can support both a single number (`10`) and a range (`10-20`)  in such a CLI design. That would make this change superfluous indeed

**Comment by jestabro** [comment]
The current thinking is to address the generalization of exit codes in a separate PR, and merge only the extensions to numeric, inversion and range values, in this PR.

**Comment by c-po** [comment]
@dmbaturin can we merge it so I can deprecate and remove the Python based range validator because of it's slownes?

**Pull request merged**